include/ensmallen_bits/nsga2/nsga2.hpp

```diff
@@ -0,0 +1,266 @@
+/**
+ * @file nsga2.hpp
+ * @author Sayan Goswami
+ *
+ * NSGA-II is a multi-objective optimization algorithm, widely used in
+ * many real-world applications. NSGA-II genererates offsprings using
+ * crossover and mutation and then selects the next generation according
+ * to nondominated-sorting and crowding distance comparison.
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef ENSMALLEN_NSGA2_NSGA2_HPP
+#define ENSMALLEN_NSGA2_NSGA2_HPP
+
+namespace ens {
+
+/**
+ * NSGA-II (Non-dominated Sorting Genetic Algorithm - II) is a multi-objective
+ * optimization algorithm. This class implements the NSGA-II algorithm.
+ *
+ * The algorithm works by generating a candidate population from a fixed
+ * starting point. At each stage of optimization, a new population of children
+ * is generated. This new population alongwith its predecessor is sorted using
+ * non-domination as the metric. Following this, the population is further
+ * segregated in fronts. A new population is generated from these fronts having
+ * size equal to that of the starting population.
+ *
+ * During evolution, two parents are randomly chosen using binary tournament
+ * selection. A pair of children are generated by crossing over these two
+ * candidates followed by mutation.
+ *
+ * The best front (Pareto optimal) is returned by the Optimize() method.
+ *
+ * For more information, see the following:
+ *
+ * @code
+ * @article{10.1109/4235.996017,
+ *   author = {Deb, K. and Pratap, A. and Agarwal, S. and Meyarivan, T.},
+ *   title = {A Fast and Elitist Multiobjective Genetic Algorithm: NSGA-II},
+ *   year = {2002},
+ *   url = {https://doi.org/10.1109/4235.996017},
+ *   journal = {Trans. Evol. Comp}}
+ * @endcode
+ *
+ * NSGA-II can optimize arbitrary multi-objective functions.  For more details,
+ * see the documentation on function types included with this distribution or
+ * on the ensmallen website.
+ */
+class NSGA2 {
+ public:
+  /**
+   * Constructor for the NSGA2 optimizer
+   *
+   * The default values provided over here are not necessarily suitable for a
+   * given function. Therefore it is highly recommended to adjust the
+   * parameters according to the problem.
+   *
+   * @param populationSize The number of candidates in the population.
```

Comment:
The NSGA implementation by the author mentions that it needs to be multiple of 4. See [this](http://www.iitk.ac.in/kangal/codes/nsga2/nsga2-gnuplot-v1.1.6.tar.gz) archive. The concerned filename is `nsga2r.c`.

Reply:
Do you think this is implementation specific?.  I'm not really seeing where in your code this is necessary, or why it might be the case. If it is required, we should check for it and terminate if it isn't a multiple of four.

Reply:
`> 4` is required for the generation of Pareto fronts (in crowding distance assignment) but being a multiple of 4 may be implementation-specific. I have added checks in the Optimizer for these two conditions.